StandardLightVisualiser : scale/radius/portal fixes and de-clutter

themissingcow opened the pull request:
### Fixes (relative to 0.56.0.0b2) 
 - Missing Arnold `spot_light` radius representation.
 - Incorrect interaction with `radius` and visualiser scale for Point/Photometric lights.
 - Fixes incorrect IES profile visualisation orientation.
 - Removes `quad_light` portal rendering artefacts with non unit square portals.

### Improvements (relative to 0.56.0.0b2)
 - Declutters `spot_light` cones (removes inner spokes from cone visualisation)
 - Adds radius representation for `spot_light`.
 - `quad_light` portal hatching now scales with visualisers.

![Screenshot 2020-02-14 at 13 24 04](https://user-images.githubusercontent.com/896779/74539457-f6c22100-4f35-11ea-877d-0d1d25e1f0dc.png)

![image](https://user-images.githubusercontent.com/896779/74367942-9f537200-4dca-11ea-8342-65f19999f9b6.png)

danieldresser-ie commented:
This definitely seems wrong.  "lensRadius" was working correctly before - it affects the shape of the cone of light where the spot is visible from.  "radius" does not affect the shape of the cone of light, but does affect the visual size of the source of the spot light as it appears in specular reflections.

Previously at ImageEngine, we scaled the color indicator on the light to indicate the "radius".  The artists I've talked to don't actually think is necessarily all that useful though ... it's maybe not a good idea, but I just mentioned it as an option.  It definitely shouldn't impact how lensRadius works though.

danieldresser-ie commented:
Thinking about it ... if we were going to visualise "radius", it would be appropriate to use something like the yellow circle around a point light, but we would probably want something less visually heavy.  Maybe just two 60 degree arcs, 180 degrees apart, cut out of a circle like the point light, except probably oriented forward?

Or maybe we don't need to visualise radius.

danieldresser-ie commented:
OK, turns out Greg Massie does feel like we should definitely have some visualisation of the radius on a spot light.  He voted for a circle like the spot light, but aligned to the direction of the light.

He felt that just showing two arcs out a circle instead of the full thing could be confusing, I still think it could look nice if tuned properly, but either way, it's probably worth having.

Other than that, one other nitpick that isn't at all urgent, but it would be nice if `spacing` and `fw` in quadPortal were set relative to the visualiser scale, rather than being hardcoded.  Currently portals really don't look nice in very large or small scenes.

themissingcow commented:
> This definitely seems wrong. "lensRadius" was working correctly before - it affects the shape of the cone of light where the spot is visible from. "radius" does not affect the shape of the cone of light, but does affect the visual size of the source of the spot light as it appears in specular reflections.

Sorry I'd totally missed that there was two (which is why I was sure it was working in the first place!) Ill put that bit back, and add in a proper visualisation for radius.

> He felt that just showing two arcs out a circle instead of the full thing could be confusing, I still think it could look nice if tuned properly, but either way, it's probably worth having.

I'll have a go at something and share later.

> Other than that, one other nitpick that isn't at all urgent, but it would be nice if spacing and fw in quadPortal were set relative to the visualiser scale, rather than being hardcoded. Currently portals really don't look nice in very large or small scenes.

Ill look into that, thanks.

themissingcow commented:
@danieldresser-ie @gregmassievfx @murrays-ie

A potential version of a radius visualisation for spots:

![image](https://user-images.githubusercontent.com/896779/74433807-091a5d00-4e59-11ea-8e2f-14e7dee5d800.png)

Im not sold on it, but Its similar to the photometric radius then. It also means that if `lens_radius` is set to the same as `radius` you can still see it.  We can try it as a single circle as per Daniels idea too (camera facing?). I noticed that Arnold suggests it behaves as a spherical source, which is why I wondered if this would work, as well as the familiarity factor with other lights.

I guess a main Q would be - what sort of ranges for radius/lens_radius do you typically work with?

themissingcow commented:
@danieldresser-ie have a look at this. We can always change the spot radius visualisation to something better. I also re-wrote the `quadPortal` hatching code, so it should get rid of the irritating gremlins a the corners with non unit quads. I'd taken some shortcuts when it only had to support 2x2, which didn't work well with other sizes.

danieldresser commented:
A quick skim all looks good.  I might play myself with some of the options for spot light radius vis, and survey some IE lighters.

The one other thing is the env light tint parameter we talked about on my closed PR, but that's simple enough I could add it myself if needed.

themissingcow commented:
> A quick skim all looks good. I might play myself with some of the options for spot light radius vis, and survey some IE lighters.

Sounds good, thanks! 

> The one other thing is the env light tint parameter we talked about on my closed PR, but that's simple enough I could add it myself if needed.

Do you  just have tint on env lights or on the other area lights too? I'd made a start on it already if you dont need it today...

andrewkaufman commented:
The tint is on area lights too

themissingcow commented:
> The tint is on area lights too

Thanks Andrew, @danieldresser-ie - this is as far as I got before end-of-play today, feel free to finish it off if you need it sooner https://github.com/themissingcow/gaffer/tree/visTint.

themissingcow commented:
@danieldresser-ie I've updated this now with the cone de-clutter and improved radius indicator. It also exposes `sphereWireframe` for you if you want to use it for a radius visualisation too.

danieldresser-ie commented:
OK, this is all looking pretty good to me now.  Folks seem happy with the new radius indicator, and the new penumbra indicator feels less intrusive.

Only thing I would say about the new radius indicator - philosophically, I feel like basically nothing in the visualisers should be hardcoded in world units, it should either be relative to the size of the lights or to the visualiserScale.  Based on that philosophy, the center of the sphereWireframe call for the radius should be either `V3f( 0.0, 0.0, 0.1 * visualiserScale )` or `V3f( 0.0, 0.0, 0.1 * radius )` instead of just `V3f( 0.0, 0.0, 0.1 )`.

One other thing I brought up before:  the thickness of the inner vs outer penumbra line.  I'm not sure how I saw the outer line being wider - in the absence of being able to reproduce it, lets assume I had done something stupid in my test scene, like accidentally dragging penumbra to be negative.  Playing with it now, I'm actually totally happy with how it looks.  I do find it a bit confusing how it's achieved in code:  the outer penumbra indicator is specified as `IECoreGL::CurvesPrimitive::GLLineWidth( 1.0f * lineWidthScale )` whereas the inner penumbra indicator is specified as `addWireframeCurveState( group.get(), lineWidthScale )`.  This produces a thicker inner cone because addWireframeCurveState contains an internal `* 2` on the width, which I don't find obvious from looking at the code, but does seem like the right result visually.

Anyway, thanks for jumping on these little changes.  I'm now feeling pretty good about the IE lights looking reasonable in Gaffer 0.56, and being a little bit more in line with how the Gaffer stuff is working.

danieldresser-ie commented:
I should add, aside from the minor nitpicks ( none of which actually need addressing at the moment ), it would be great to get this merged ASAP, because now the version of the IE tools that I'm working on is dependent on this version, making it a bit tricky keeping everything in sync.

themissingcow commented:
> OK, this is all looking pretty good to me now. Folks seem happy with the new radius indicator, and the new penumbra indicator feels less intrusive.

Epic, thanks for taking all the time to run it by folks. 👍 

> Only thing I would say about the new radius indicator - philosophically, I feel like basically nothing in the visualisers should be hardcoded in world units, it should either be relative to the size of the lights or to the visualiserScale. Based on that philosophy, the center of the sphereWireframe call for the radius should be either `V3f( 0.0, 0.0, 0.1 * visualiserScale )` or `V3f( 0.0, 0.0, 0.1 * radius )` instead of just `V3f( 0.0, 0.0, 0.1 )`.

100% - I'd forgotten this was in `fixedScaleOrnaments`(as opposed to `ornaments`), fixed in https://github.com/GafferHQ/gaffer/pull/3616/commits/b941a417e5850a871f0c71b682cd337e41dc714c.

>  I'm actually totally happy with how it looks. I do find it a bit confusing how it's achieved in code: the outer penumbra indicator is specified as `IECoreGL::CurvesPrimitive::GLLineWidth( 1.0f * lineWidthScale )` whereas the inner penumbra indicator is specified as `addWireframeCurveState( group.get(), lineWidthScale )`.

Sorry, I think this just snuck in over the course of the various rejigs that have gone on in here. Fixed in https://github.com/GafferHQ/gaffer/pull/3616/commits/c9db25117ae8ee5471f36d2b0ab91e9e2768a955. 

> Anyway, thanks for jumping on these little changes. I'm now feeling pretty good about the IE lights looking reasonable in Gaffer 0.56, and being a little bit more in line with how the Gaffer stuff is working.

Thats great news. Thanks for all the patience and help getting this in shape.

themissingcow commented:
@johnhaddon Be great if we could get this merged today for Daniel.

johnhaddon commented on src/GafferSceneUI/StandardLightVisualiser.cpp:
```diff
@@ -451,13 +451,14 @@ Visualisations StandardLightVisualiser::visualise( const IECore::InternedString
 		float innerAngle, outerAngle, radius, lensRadius;
 		spotlightParameters( attributeName, shaderNetwork, innerAngle, outerAngle, radius, lensRadius );
 		boundOrnaments->addChild( const_pointer_cast<IECoreGL::Renderable>( spotlightCone( innerAngle, outerAngle, lensRadius / visualiserScale, 1.0f, 1.0f ) ) );
+		fixedScaleOrnaments->addChild( const_pointer_cast<IECoreGL::Renderable>( sphereWireframe( radius, Vec3<bool>( false, false, true ), 0.5f, V3f( 0.0, 0.0, 0.1 * visualiserScale ) ) ) );
```
I'm not a fan of the radius visualisation being moved according to visualiser scale. Doesn't the radius represent the _geometry_ that shadow rays get fired at, and isn't moving that around misleading? I guess the argument is that we need to keep it out of the way of the colour indicator?

themissingcow replied:
Its all a bit abstract, for Arnold, its described as:

>  With a non-zero value for radius, then the light source will behave more like a spherical source with the radius specified (although again with the light emitted being constrained to a cone)

In practice, its some kind of hybrid, you can make the radius larger than some object in your scene, and that object still receives shadowing...

themissingcow replied:
>  I guess the argument is that we need to keep it out of the way of the colour indicator? 

I think the argument is that unless some part of the visualiser has well-represented world meaning (eg: `lens_radius`, a vis at scale of 1 and 10 should look the same, just bigger.

themissingcow replied:
(In this case the offset would not have a world meaning, but the radius of it does)

johnhaddon replied:
> In practice, its some kind of hybrid, you can make the radius larger than some object in your scene, and that object still receives shadowing...

I'm not sure what this means, but it still seems that the radius is centred at `0, 0, 0` in local space, and not at some arbitrary distance behind it.

themissingcow replied:
This was why I did it originally:
![r1s1](https://user-images.githubusercontent.com/896779/74669108-689ca380-519e-11ea-97d2-b58dffbdd6dc.png)
![r10s10wo](https://user-images.githubusercontent.com/896779/74669111-69cdd080-519e-11ea-9693-6552e0a1432a.png)
![r10s10](https://user-images.githubusercontent.com/896779/74669110-69cdd080-519e-11ea-9fbf-5561aa74c055.png)

If you put `radius` vis at `0, 0, 0` then when `lens_radius` == `radius`, as `lens_radius` (which does have absolute world meaning as it represents the visible source of light emission) is already at `0, 0, 0`, you can't see the radius visualisation at all. Daniel attempted to address this by adding chevrons onto the side, which we could add to this, but we generally preferred the simplicity and consistency without.

themissingcow replied:
> > In practice, its some kind of hybrid, you can make the radius larger than some object in your scene, and that object still receives shadowing...
> 
> I'm not sure what this means, but it still seems that the radius is centred at `0, 0, 0` in local space, and not at some arbitrary distance behind it.

![image](https://user-images.githubusercontent.com/896779/74669782-c67dbb00-519f-11ea-8088-dc58e503ef4e.png)

The cube is 2 units from the light, of `radius` 5.

themissingcow replied:
Other lights, such as the `photometic_light` and `point_light` don't emit inside their radius:

![image](https://user-images.githubusercontent.com/896779/74669948-19f00900-51a0-11ea-8f60-a9f3ad517a92.png) ![image](https://user-images.githubusercontent.com/896779/74670017-3724d780-51a0-11ea-91b3-92544fc260f5.png)

themissingcow replied:
How about this as a compromise: https://github.com/GafferHQ/gaffer/pull/3616/commits/ebed8d3886c577226e5f395f057e90e8ca7114a2

themissingcow replied:
`<chatted>` and decided to just leave it as the original PR for now and see how it goes.

johnhaddon replied:
Summary of offline conversation :

- b8e9af5 has the downside that the radius intersects with the colour indicator.
- We think there is a sphere centered at 0 and that shadow rays are thrown at it, but it's confusing when the radius gets big because the shadow rays end up diverging a long way from the light vector.
- I quite like the idea of using a non-offset sphere visualisation, which would still intersect the colour indicator but would be clearer due to having rings in two axes.
- The current offset disk has been run past quite a few people and approved though, so I should let it go. We can revisit later if necessary.

danieldresser replied:
> We think there is a sphere centered at 0 and that shadow rays are thrown at it, but it's confusing when the radius gets big because the shadow rays end up diverging a long way from the light vector.

Arnold is inconsistent about whether the source of spotlight is a circle or a sphere, when it comes to the penumbra, some calculations use one, some use the other - this actually causes ugly artifacts in production, which we've been asking them to fix.  Not particularly relevant, except that trying to measure Arnold's behaviour to decide one way or the other.

> I quite like the idea of using a non-offset sphere visualisation, which would still intersect the colour indicator but would be clearer due to having rings in two axes.
If this is refering to Tom's original one with one circle perpendicular to the direction of the light, running straight through the cone, it was pretty unpopular at IE.  Maybe just leave 20 degrees worth of the perpendicular circle where it intersects the plane of the light?  Would that be a form of chevron that would bother Tom less?

> The current offset disk has been run past quite a few people and approved though, so I should let it go. We can revisit later if necessary.

Sounds like the right call.

themissingcow commented on src/GafferSceneUI/StandardLightVisualiser.cpp:
```diff
@@ -996,73 +996,81 @@ IECoreGL::ConstRenderablePtr StandardLightVisualiser::quadPortal( const V2f &siz
-	static const float spacing = 0.05f;
-	// size.x of the shaded frame area
-	static const float fw = 0.25f;
+	const float spacing = 0.05f * visualiserScale;
```
The opposite is bad too though, as at 100x, the lines are 0.05 real world units apart and it's just solid black... Maybe we just make the spacing a proportion of the shortest side.

johnhaddon replied:
I don't follow. If I scale my light 100x (scaling is the only thing you can do in Gaffer at present, we don't have width/height controls) I get exactly the same number of lines as I did at 1x scale, and it all looks great. It's only when I mess with _visualiser_ scale that it all goes wrong.

themissingcow replied:
How about this as a compromise: https://github.com/GafferHQ/gaffer/pull/3616/commits/d6f5ef6cb5c0bd2c4c1fbaba8504afecba6954a5

In `StandardLightVisualiser` we keep the old behaviour (as we only have a fixed quad size, so people _have_ to scale the light). Then @danieldresser-ie can call it with `visualiserScale` for his lights, where they have `width` and `height`?

themissingcow replied:
> I don't follow. If I scale my light 100x (scaling is the only thing you can do in Gaffer at present, we don't have width/height controls) I get exactly the same number of lines as I did at 1x scale, and it all looks great. It's only when I mess with _visualiser_ scale that it all goes wrong.

Did you see Daniels comment earlier that it looks terrible at IE as they have big lights, so it's a wall of black hatching?

johnhaddon replied:
> Did you see Daniels comment earlier that it looks terrible at IE as they have big lights, so it's a wall of black hatching?

Yes, I assumed that was your motivation for breaking the Cinesite hatching ;)

> How about this as a compromise: d6f5ef6

Seems OK, but I presume Daniel will end up doing something like `hatchingScale = 1 / max( width, height )`, in which case maybe we should just do that automatically internally? I presume you were thinking along similar lines when you said "Maybe we just make the spacing a proportion of the shortest side."

themissingcow replied:
I'd assumed he'll just do `hatchingScale = visualiserScale`, as that's what he wanted originally?

themissingcow replied:
I thought of one potential benefit of keeping it de-coupled from w/h - when we do support `width`/`height` editing, any interactive manipulations might be less distracting (though still not ideal) than with something based on the dimensions? I guess ideally, to make that the least eye melting it could be, we'd quantise the hatching to a grid rather than bottom left.

johnhaddon replied:
> I'd assumed he'll just do hatchingScale = visualiserScale, as that's what he wanted originally?

But surely that only fixes the high hatching density for large width and/or height if the visualiser scale is adjusted to compensate?

> I thought of one potential benefit of keeping it de-coupled from w/h - when we do support width/height editing, any interactive manipulations might be less distracting (though still not ideal) than with something based on the dimensions? I guess ideally, to make that the least eye melting it could be, we'd quantise the hatching to a grid rather than bottom left.

Agreed. Perhaps it'd also be nice to transition between hatching densities by introducing new lines smoothly in between the existing ones. But I don't think any of this is helped by pushing a `scale` parameter into `quadPortal()` now - then all a client can do is make some poor approximation of the ideal behaviour, and we're a bit more constrained to implement that ideal behaviour later.

johnhaddon replied:
> and we're a bit more constrained to implement that ideal behaviour later.

Actually, this is overstating things. I don't see any harm in a scale parameter if it helps Daniel in the short term.  d6f5ef6 will do for me...

themissingcow replied:
> But surely that only fixes the high hatching density for large width and/or height if the visualiser scale is adjusted to compensate?

Yeah, I guess related to the other point above though, when I was testing it as I was writing, there is something nice about looking at the scene with consistent hatch spacing across the lights in a scene. It bugs me that they scale with the light in stock Gaffer, but there's nothing we can do about that. 

One of the requests we had originally that made us decouple ornaments from geometry scaling was that uniform scale for common elements helps you understand depth better on a flat screen.  It felt like the vis scale linked approach helped with that, as no matter the size of your lights (at IE, for now) you'd be able to grok their relative position/sizes better.

I think for a scene where the lighter (or facility/show) has set a sensible visualiser scale for the working world scale, uniform hatching wins for me from that perspective. If you've not adjusted vis scale accordingly in the first, then so many other things look bad.

Im happy either way though if you and @danieldresser-ie want to work out what we want to go with.

johnhaddon replied:
Thanks for the explanation. I think d6f5ef6 is the best we can do for now then - doesn't break the Gaffer visualisations, and allows the IELight visualisations to be a bit nicer. We can revisit when we start using width/height rather than scale in Gaffer.

danieldresser replied:
> It bugs me that they scale with the light in stock Gaffer, but there's nothing we can do about that.

Any reason we can't divide off the world space scale when determining the hatching?  Or do visualisers not get recomputed when they're scaled?  Hmm, it's too bad, making the hatching independent of the size of the light seems like the obviously correct answer, to keep all the lights nicely matching each other, and have everything adjust when a global visualiserScale for the scene is dialed in.

I guess I'm just lucky that I've already been trying to discourage anyone from scaling lights at IE.

johnhaddon replied:
> Any reason we can't divide off the world space scale when determining the hatching? Or do visualisers not get recomputed when they're scaled?

Yep, visualisers are generated before scaling. There are a couple of options though, one of which Tom and I discussed yesterday :

- Use a shader to generate the hatching on the fly
- Use a custom `IECoreGL::Renderable` subclass to generate the hatching on the fly

We plan to revisit this at some point, but I don't think it's worth holding things up for at present.

themissingcow commented on src/GafferSceneUI/StandardLightVisualiser.cpp:
```diff
@@ -643,6 +643,10 @@ void StandardLightVisualiser::spotlightParameters( const InternedString &attribu
 		lensRadius = parameter<float>( metadataTarget, shaderParameters, "lensRadiusParameter", 0.0f );
 	}
+
+	radius = 0.0f;
```
A very good point indeed, thanks! https://github.com/GafferHQ/gaffer/pull/3616/commits/1a44c76c3935639a3506733863c82d9690e04685

themissingcow commented:
@johnhaddon As per off-line chat (via the internet 🙃) have updated https://github.com/GafferHQ/gaffer/pull/3616/commits/c8976bacc150baeefd09207389bcdac8c9c896d0 to revert stock-gaffer behaviour. And left the `radius` offset in for now. The other code review comments are fixed in https://github.com/GafferHQ/gaffer/pull/3616/commits/281087366c0eea7da22c3dba1f74fb8544306606 too.

johnhaddon commented:
Thanks Tom. @danieldresser-ie, I intend to have gone home before CI runs through, so please feel free to merge once Azure gives the green light.

danieldresser commented:
I wasn't at work today, and there was a conflict, so I haven't hit merge, but it sounds like the concensus is that this is all good enough for now, so we should merge and move forward.

johnhaddon commented:
> I wasn't at work today, and there was a conflict, so I haven't hit merge, but it sounds like the concensus is that this is all good enough for now, so we should merge and move forward.

Could you resolve and merge please @themissingcow?

themissingcow commented:
> > I wasn't at work today, and there was a conflict, so I haven't hit merge, but it sounds like the concensus is that this is all good enough for now, so we should merge and move forward.
> 
> Could you resolve and merge please @themissingcow?

Done that there ting.
